### PR TITLE
Add colors and total problems to the output

### PR DESCRIPTION
**Pull request by wesleimp**
Add colors to the output and total problems for a better visual experience.
* `file:line` - italic
* `checkname` - bold red

The flag `--no-color` turns off the feature, leaving only the plain text

image

#### ✔ Checklist:

- [x] This PR has been added to [CHANGELOG.md](https://github.com/dotenv-linter/dotenv-linter/blob/master/CHANGELOG.md) (at the top of the list);
- [x] Tests for the changes have been added (for bug fixes / features);
- [ ] Docs have been added / updated (for bug fixes / features).

### CODE REVIEW

**Comment by wesleimp** [comment]
> Hey thank you for your PR rocket
> In general a very cool idea, I like to see how many errors are there in total but I think what would be really cool would be to have this option on as default but have an argument to switch colouring off.
> That way it would be easier to process for non-humans i.e. extracting information, act on specific errors etc. or if you redirected the output into a file I think all colours and formatting will be in there ascii encoded so that the content would be hard to read.
> 
> Anyway, not wanting to make your change bad :) It's great! And I think this suggestion could be even handled in a separate PR.

@mstruebing I got your point and I like it. I'll wait for the other reviewers and do that after merge.

What do you think?

**Comment by wesleimp** [comment]
> Hey thank you for your PR rocket
> In general a very cool idea, I like to see how many errors are there in total but I think what would be really cool would be to have this option on as default but have an argument to switch colouring off.
> That way it would be easier to process for non-humans i.e. extracting information, act on specific errors etc. or if you redirected the output into a file I think all colours and formatting will be in there ascii encoded so that the content would be hard to read.
> 
> Anyway, not wanting to make your change bad :) It's great! And I think this suggestion could be even handled in a separate PR.

@mstruebing thinking about you review, maybe doesn't make sense adding this feature without one way to turns it off. So, I've added the flag `--no-color` too.

File: src/main.rs [hunk 1]
```diff
@@ -14,13 +15,46 @@ fn main() -> Result<(), Box<dyn Error>> {
...
     }
 
-    warnings.iter().for_each(|w| println!("{}", w));
-    process::exit(1);
+    print_warnings(args.is_present("no-color"), warnings);
+    process::exit(1)
+}
+
+fn print_warnings(no_color: bool, warnings: Vec<dotenv_linter::common::Warning>) {
+    if no_color {
```
Comment (DDtKey) [inline src/main.rs:28]: It seems to be better to create two separate functions for _color output_ and _no-color_. 
This will make the code much easier to read.

File: src/main.rs [hunk 2]
```diff
@@ -14,13 +15,46 @@ fn main() -> Result<(), Box<dyn Error>> {
...
+        warnings.iter().for_each(|w| {
+            println!(
+                "{}:{} {}: {}",
+                w.line.file.path.display(),
+                w.line.number,
+                w.check_name,
+                w.message
+            )
+        });
+        println!("{} problems", warnings.len());
```
Comment (DDtKey) [inline src/main.rs:38]: I doubt we need print warning count as least for `no-color`output :thinking: 
This may be redundant for CI systems and etc.

File: src/lib.rs [hunk 3]
```diff
@@ -4,13 +4,13 @@ use std::error::Error;
 use std::path::PathBuf;
 
 mod checks;
-mod common;
+pub mod common;
```
Comment (DDtKey) [inline src/lib.rs:7]: I think it would be better to use `pub(crate)` in this place.

File: src/common.rs [hunk 4]
```diff
@@ -24,8 +25,11 @@ impl fmt::Display for Warning {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         write!(
             f,
-            "{}:{} {}: {}",
-            self.line.file, self.line.number, self.check_name, self.message
+            "{}:{} {} {}",
```
Comment (DDtKey) [inline src/common.rs]: I would change that. 
Since at the moment this code has become unused and `color`/`no-color` is controlled via the argument.

Or I would have thought about how to do formatting at the `Warning` level without special output in the `main.rs`.

File: src/common.rs [hunk 5]
```diff
@@ -4,9 +4,9 @@ use std::path::PathBuf;
...
 pub struct Warning {
-    check_name: String,
-    line: LineEntry,
-    message: String,
+    pub check_name: String,
+    pub line: LineEntry,
+    pub message: String,
 }
 
 impl Warning {
```
Comment (DDtKey) [inline src/common.rs:12]: This part is out of control if we just use the output somewhere, which confuses me a little.

I looked a little at the possibilities of `colored` and it seems to me that it will be correct to use [their internal way of configuring the output](https://github.com/mackwic/colored#colorization-control) if it possible.

What do you think? :thinking:

Comment (DDtKey) [reply src/common.rs:12]: Overall, I like the current implementation :+1: 

Now the changes mostly affect only the `main.rs` file and it works obviously.

Comment (wesleimp) [reply src/common.rs:12]: Nice, I'll work on it

File: src/lib.rs (hunk 3 above)
Comment (wesleimp) [reply src/lib.rs]: @DDtKey good catch, however this code has been removed

File: src/main.rs [hunk 6]
```diff
@@ -14,13 +15,33 @@ fn main() -> Result<(), Box<dyn Error>> {
...
+}
+
+fn print_total_problems(total_problems: usize) {
+    println!();
+    println!(
+        "{}",
+        format!("{} {} problems", "\u{2717}", total_problems)
+            .red()
+            .bold()
+    );
```
Comment (wesleimp) [inline src/main.rs:43]: @DDtKey sorry, this refactor dismissed your review :grimacing:

Comment (DDtKey) [reply src/main.rs:43]: > @DDtKey sorry, this refactor dismissed your review grimacing

Good catch! 
Thanks, it's okay :relaxed:

File: src/main.rs [hunk 7]
```diff
@@ -14,13 +15,33 @@ fn main() -> Result<(), Box<dyn Error>> {
...
+
     warnings.iter().for_each(|w| println!("{}", w));
-    process::exit(1);
+    if should_print_total_problems {
+        print_total_problems(warnings.len())
+    }
+    process::exit(1)
+}
+
+fn print_total_problems(total_problems: usize) {
```
Comment (mgrachev) [inline src/main.rs:36]: I've pulled out these changes into another PR and added an additional flag not to display this message.
Please, review it https://github.com/dotenv-linter/dotenv-linter/pull/242 🙂

I think we should merge my PR #242 before this one.

**Comment by mgrachev** [comment]
@wesleimp Please, resolve conflicts.

**Comment by mgrachev** [comment]
@wesleimp Please, update your PR from the master branch.

**Comment by mgrachev** [comment]
@wesleimp Gentle ping 🙂

**Comment by mgrachev** [comment]
@wesleimp Do you have any progress on this?

**Comment by mgrachev** [comment]
@wesleimp I'm sorry, but I'll close this inactive PR.
